### PR TITLE
ENH/BUG: Add No Modulus Option to SLT Calculations, .copy() Fix, and More!

### PR DESCRIPTION
**Pull request by rstoneback**
# Description

Addresses SLT upgrade request in #723, bug in #728, and more!

 - Adds support to disable constraining SLT values to [0, 24). Variable is continuous when loading multiple days.
 -  Fixes .copy(). Extends .copy() and __eq__ across Instrument classes. Tests added. 
 - Tests with __repr__()
 - Expands general Instrument tests to Instruments instantiated via `inst_module`
 - Condenses methods used by pysat testing instruments. Adds support for keyword tests to all.
 - I would appreciate feedback on formatting support function docstrings in testing.py
 - Updates the initialization keyword argument process. 
   - Always an entry for each Instrument function.
   - 'partial' function keywords retained;  eval(__repr__()) would otherwise lose those keywords
   - Switch all keyword treatment over to 'partial'. All keywords, user provided at instantiation or not, were listed in self.kwargs. This lead to a difference between actual user input and the string provided by __repr__. In some situations, the string from __repr__ and the string from eval(__repr__).__repr__ would have keywords in a different order, frustrating string comparisons, and would produce an underlying difference in `self.kwargs`.
   - Added `inst.kwargs_supported`. Includes all supported keyword and default values.
   
   -  I also updated code to always pass in the kwargs for each function when run (except list_files). This ensure liveness, if a user tweaks inst.kwargs it will be reflected in outputs. list_files doesn't work since files runs that method and the weakef that was in files back to Instrument is now gone.
       - Should I update `Files` to include a full reference to Instrument as done in orbits? **UPDATE** I went ahead and put this in and updated the tests. It is now 'live' as well. I have a weakref back to inst contained within files.inst_info. Switched orbits back to weakref.proxy as well. Working fine so far.
 ```
In [2]: eval("pysat.Instrument(platform='pysat', name='testing', inst_id='', clean_level='clean', pad=None, orbit_info={'index': 'mlt'}, inst_module=None, custom=[], **{'sim_multi_file_right': False, 'sim_multi_file_left': False, 'root_date': None, 'malformed_index': False, 'num_samples': None, 'test_load_kwrd': None, 'file_date_range': None, 'mangle_file_dates': False, 'test_init_kwrd': None})")                        

Out[2]: pysat.Instrument(platform='pysat', name='testing', inst_id='', clean_level='clean', pad=None, orbit_info={'index': 'mlt'}, inst_module=None, custom=[], **{'file_date_range': None, 'sim_multi_file_right': False, 'sim_multi_file_left': False, 'root_date': None, 'malformed_index': False, 'num_samples': None, 'test_load_kwrd': None, 'mangle_file_dates': False, 'test_init_kwrd': None})
```

 - I was able to remove the need for modifying file lists from within `init` for the testing instruments
 - Updated tests, and fixed some bugs with `remote_file_list` and `download` related to `mangled_file_list'
 - Removed deprecated 'modify' keywords in orbits test code
 - Added check for NoneType within `to_netcdf4`

These changes may appear unrelated however the core, the SLT upgrade, plus the inst_module bug, are combined since this branch is currently being used by some local software that needs both. The remaining expansion of commits all came up as I was working on the core stuff.

## Type of change

Please delete options that are not relevant.

- New feature (non-breaking change which adds functionality)
- Bug-fix

# How Has This Been Tested?

UnitTests plus plots!

# Checklist:

- [ ] Make sure you are merging into the ``develop`` (not ``main``) branch
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] Add a note to ``CHANGELOG.md``, summarizing the changes

### CODE REVIEW

**Comment by rstoneback** [comment]
I applied the code to some of the testing instruments.

```
In [39]: py_inst = pysat.Instrument(platform='pysat', name=name) 
    ...: py_inst.load(date=dt.datetime(2010, 1, 1), end_date=dt.datetime(2010, 1, 3)) 
    ...: coords.calc_solar_local_time(py_inst, lon_name="longitude", 
    ...:                                       slt_name='slt', apply_modulus=False)                                                                                                                                          
```
multi_day_slt_testing
multi_day_slt_testing2d_xarray
multi_day_slt_testmodel

**Comment by rstoneback** [comment]
I expanded testing to include a pysat.Instrument instantiated via a module and a few more errors popped up, expanding the range of commits.

**Comment by rstoneback** [comment]
I made a new function over in methods/testing.py so all test instruments could support the same changes to file lists (part of unit testing). As part of this, I also moved keywords for clarity, then realized 'init' was not a supported function for keywords. I modified that, added a test, and here we are. While all of these commits are actually centered around getting the SLT and other things going, I'd like to close this out before I get further afield.

File: pysat/instruments/pysat_testing.py [hunk 1]
```diff
@@ -101,9 +92,8 @@ def preprocess(self):
 
 
 def load(fnames, tag=None, inst_id=None, sim_multi_file_right=False,
-         sim_multi_file_left=False, root_date=None, file_date_range=None,
-         malformed_index=False, mangle_file_dates=False, num_samples=None,
-         multi_file_day=False):
+         sim_multi_file_left=False, root_date=None, malformed_index=False,
+         num_samples=None, multi_file_day=False):
```
Comment (rstoneback) [inline pysat/instruments/pysat_testing.py:96]: the `multi_file_day` keyword doesn't go to anything that I can see. I wasn't directly in on the recent-is `multi_file_day` change, perhaps this is a holdover?

Comment (aburrell) [reply pysat/instruments/pysat_testing.py]: That sounds highly possible.

File: docs/new_instrument.rst [hunk 2]
```diff
@@ -36,7 +36,7 @@ part of pysat's user instrument registry using the following syntax:
   registry.register_from_module(my_package.instruments)
 
 After registry the instrument module name is stored in the user's home
-directory under :code:`~.pysat/user_modules.txt`. The instrument may then
+directory under :code:`~.pysat`. The instrument may then
```
Comment (aburrell) [inline docs/new_instrument.rst:39]:
```suggestion
directory in a hidden directory named :code:`~.pysat`. The instrument may then
```

File: docs/new_instrument.rst [hunk 3]
```diff
@@ -433,33 +433,44 @@ allowing them to be stored as a pandas DataFrame. Setting this attribute to
...
-Custom Keywords in load Method
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Custom Keywords in Support Methods
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 If provided, pysat supports the definition and use of keywords for an
-instrument module so that users may trigger optional features. All custom
-keywords for an instrument module must be defined in the ``load`` method.
-
+instrument module so that users may trigger optional features. A custom
```
Comment (aburrell) [inline docs/new_instrument.rst:440]:
```suggestion
instrument module so that users may define their preferred default values. A custom
```

File: pysat/_files.py [hunk 4]
```diff
@@ -266,6 +268,60 @@ def __str__(self):
...
+
+        Parameters
+        ----------
+        other : any
+            Other object to compare for equality
+
+        Returns
+        -------
+        bool
+            True if objects are identical
```
Comment (aburrell) [inline pysat/_files.py:282]:
```suggestion
            True if objects are identical, False if they are not
```

File: pysat/_files.py [hunk 5]
```diff
@@ -266,6 +268,60 @@ def __str__(self):
 
         return output_str
 
+    def __eq__(self, other):
```
Comment (aburrell) [inline pysat/_files.py:271]: This method needs comments to clarify what each bit of code is doing.

File: pysat/_files.py [hunk 6]
```diff
@@ -545,6 +601,25 @@ def _remove_data_dir_path(self, file_series=None):
...
+        """Provide a deep copy of object
+
+        Returns
+        -------
+        Files class instance
+            Copy of self
+
+        """
+        saved_info = self.inst_info
+        self.inst_info = None
```
Comment (aburrell) [inline pysat/_files.py:614]: Why is this necessary?  Recommend adding a comment to clarify.

**Review by aburrell — COMMENTED** [review]
Left comments on Files, it looks like they're relevant throughout.  Bit of a shame so many things are combined in the PR, but 🐭

File: pysat/_instrument.py [hunk 7]
```diff
@@ -456,11 +468,83 @@ def __init__(self, platform=None, name=None, tag=None, inst_id=None,
...
+
+        Parameters
+        ----------
+        other : any
+            Other object to compare for equality
+
+        Returns
+        -------
+        bool
+            True if objects are identical
```
Comment (rstoneback) [inline pysat/_instrument.py:487]:
```suggestion
            True if objects are identical, False if they are not.
```

File: pysat/_orbits.py [hunk 8]
```diff
@@ -155,6 +158,62 @@ def __str__(self):
...
+
+        Parameters
+        ----------
+        other : any
+            Other object to compare for equality
+
+        Returns
+        -------
+        bool
+            True if objects are identical
```
Comment (rstoneback) [inline pysat/_orbits.py:172]:
```suggestion
            True if objects are identical, False if they are not
```

**Comment by rstoneback** [comment]
>  Bit of a shame so many things are combined in the PR, but 🐭

Yeah.... apparently once I get going I get going. One of these days I'll manage to get a fully proper pull together.

File: pysat/tests/test_orbits.py [hunk 9]
```diff
@@ -3,6 +3,7 @@
 import numpy as np
 
 import pandas as pds
+from pandas import Timedelta # noqa
```
Comment (aburrell) [inline pysat/tests/test_orbits.py:6]: Why is this line in here and why is flake8 ignored?  Why not just use `pds.Timedelta`?

File: pysat/tests/test_orbits.py [hunk 10]
```diff
@@ -239,6 +240,65 @@ def teardown(self):
...
+    def test_equality_with_copy(self):
+        """Test that copy is the same as original"""
+        self.out = self.testInst.orbits.copy()
+        assert self.out == self.testInst.orbits
+
+    def test_equality_with_data_with_copy(self):
+        """Test that copy is the same as original"""
+        # Load data
+        self.testInst.load(date=self.stime)
+        # Load up an orbit
```
Comment (aburrell) [inline pysat/tests/test_orbits.py:252]:
```suggestion
        self.testInst.load(date=self.stime)

        # Load up an orbit
```

Comment (aburrell) [reply pysat/tests/test_orbits.py:252]: Here and elsewhere, to make the code easier to read.

File: pysat/tests/test_utils_coords.py [hunk 11]
```diff
@@ -43,6 +43,7 @@ def setup(self):
         """Runs after every method to clean up previous testing."""
         self.py_inst = None
         self.inst_time = dt.datetime(2009, 1, 1)
+        self.inst_2_time = dt.datetime(2009, 1, 3)
```
Comment (aburrell) [inline pysat/tests/test_utils_coords.py:46]:
```suggestion
        self.inst_time_2 = dt.datetime(2009, 1, 3)
```

Because it's the second instrument time, not the second instrument's time.

File: pysat/utils/coords.py [hunk 12]
```diff
@@ -89,6 +90,9 @@ def calc_solar_local_time(inst, lon_name=None, slt_name='slt'):
         name of the longtiude data key (assumes data are in degrees)
     slt_name : string
         name of the output solar local time data key (default='slt')
+    apply_modulus : bool
+        If True, slt values are confined to [0, 24) via a modulus operation.
+        If False, no modulus applied. (default=True)
```
Comment (aburrell) [inline pysat/utils/coords.py:95]:
```suggestion
        If True, SLT values are confined to [0, 24), if False they may be positive or negative based on the value of their universal time relative to that of the first point included in the Instrument object. (default=True)
```

With this, it seems like there should also be an option to supply a reference time that defaults to the first time value.

File: pysat/utils/coords.py [hunk 13]
```diff
@@ -103,12 +107,18 @@ def calc_solar_local_time(inst, lon_name=None, slt_name='slt'):
 
     # Convert from numpy epoch nanoseconds to UT seconds of day
     ut_hr = list()
+    first_time = inst.index.values[:1].astype(int)[0]
     for nptime in inst.index.values.astype(int):
+        # Account for any difference in days across loaded data (multi-day span)
+        day_diff = np.floor((nptime - first_time) * 1.0e-9 / 86400.)
+
```
Comment (aburrell) [inline pysat/utils/coords.py:114]: Wouldn't it be quicker to only do this if needed? You could just set it to zero if not needed.

File: pysat/utils/coords.py [hunk 14]
```diff
@@ -147,7 +157,13 @@ def calc_solar_local_time(inst, lon_name=None, slt_name='slt'):
                 slt[i] = hr + inst[lon_name] / 15.0
 
     # Ensure that solar local time falls between 0 and 24 hours
```
Comment (aburrell) [inline pysat/utils/coords.py:159]: Fix comment to be accurate

File: pysat/utils/coords.py [hunk 15]
```diff
@@ -147,7 +157,13 @@ def calc_solar_local_time(inst, lon_name=None, slt_name='slt'):
...
-    slt = np.mod(slt, 24.0)
+    if apply_modulus:
+        slt = np.mod(slt, 24.0)
+        min_val = 0.0
+        max_val = 0.0
+    else:
+        min_val = np.nan
+        max_val = np.nan
```
Comment (aburrell) [inline pysat/utils/coords.py:166]: Technically, shouldn't these be +/- infinity?  Not invested either way!

**Review by aburrell — COMMENTED** [review]
Again, several comments (especially in Files equality) apply to other related routines.  I recommend this PR not be merged without approval from both myself and @jklenzing, because the scope is large and I am likely to miss something.

File: pysat/tests/test_orbits.py (hunk 9 above)
Comment (jklenzing) [reply pysat/tests/test_orbits.py:6]: Looking through the file, timedelta is only used as `dt.timedelta`.  This could probably be deleted.

**Review by jklenzing — CHANGES_REQUESTED** [review]

**Comment by rstoneback** [comment]
Thanks for the feedback. I'm taking a few days off and will get back to this early next week.

File: pysat/tests/test_instrument.py [hunk 16]
```diff
@@ -26,12 +34,14 @@
 # -----------------------------------------------------------------------------
 class TestBasics():
     def setup(self):
+        global testing_kwrds
```
[the remark this answers is not in the document]
Comment (aburrell) [reply pysat/tests/test_instrument.py:37]: I thought so too, but @rstoneback requested this a few pulls ago.

File: pysat/tests/test_instrument.py [hunk 17]
```diff
@@ -593,6 +604,106 @@ def test_filtered_date_attribute(self):
...
+    #
+    # Test copy method
+    #
+    # -------------------------------------------------------------------------
+
+    def test_copy(self):
+        """Test .copy()"""
+        inst_copy = self.testInst.copy()
+        assert inst_copy == self.testInst
+        return
```
[the remark this answers is not in the document]
Comment (aburrell) [reply pysat/tests/test_instrument.py:683]: I think that's me.  I always look for a `return` to know that a method or function has ended.  We can discuss this as part of pysat style next meeting, if you like.

**Comment by aburrell** [comment]
Looking over this, all except for the copy issue seem to be enhancements.  There are also some issues it addresses (simplifying test instruments, adding more equality functions) that were previously put off until 3.1.0+.  Should this PR be rebased to the develop-3 branch and a separate issue put into the RC to address the copy issue?

File: pysat/tests/test_instrument.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (jklenzing) [reply pysat/tests/test_instrument.py:683]: I'm fine either way, just want to know what to use for consistency

**Comment by rstoneback** [comment]
> Looking over this, all except for the copy issue seem to be enhancements. There are also some issues it addresses (simplifying test instruments, adding more equality functions) that were previously put off until 3.1.0+. Should this PR be rebased to the develop-3 branch and a separate issue put into the RC to address the copy issue?

The pull went to the code areas it did because of bugs and testing. Then tests needed to be there to confirm the bug was gone, which then required things like equality methods, copy, etc, testing of which revealed its own stuff.  I'm not sure how to just pull in the 'copy issue' without the rest since other than the SLT issue the code areas are linked.

I was unaware some of these changes were listed as issues for 3.1. I don't support a delay.

File: pysat/_files.py [hunk 18]
```diff
@@ -266,6 +268,81 @@ def __str__(self):
...
+        checks = []
+        item_check = []
+        for item in self.__dict__:
+            item_check.append(item)
+            # Confirm each object has the same items
+            if item in other.__dict__:
+                # Define default comparison. Series comparisons can error
+                # for simply being different (files, _*_file_list).
+                # 'inst_info' contains a weakref back to Instrument that
+                # requires a different check.
```
[the remark this answers is not in the document]
Comment (aburrell) [reply pysat/_files.py]: Yes, I know that.  Why is it important to say that in this comment?  Why is it not enough to simply use that test?

**Comment by rstoneback** [comment]
I agree there are a bunch of changes. I didn't intend for the pull to become so expansive. While the intent was bug fixes, I also agree that it stretches the definition and is a blurry combo of enhancements. I also agree that this level of change perhaps shouldn't be going in at the RC level. If we leave it out though I can't help feel that we are choosing to leave bugs in.
Perhaps this is large enough to also talk Friday? Have another meeting to run to.

File: pysat/tests/test_instrument.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (aburrell) [reply pysat/tests/test_instrument.py:694]: We can clean that up alongside the docstrings in the future.

**Comment by aburrell** [comment]
> Perhaps this is large enough to also talk Friday?

Yes, absolutely.

File: pysat/_files.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (rstoneback) [reply pysat/_files.py]: I am surprised that a Series comparison would error rather than return False. I thought I was adding clarity but I can remove the comment.

File: pysat/tests/test_orbits.py (hunk 9 above)
Comment (aburrell) [reply pysat/tests/test_orbits.py:7]: Curious, why does `__repr__` need this? How can it need something that isn't called? This seems very un-pythonic and I am confused. 🐒

File: pysat/utils/coords.py [hunk 19]
```diff
@@ -101,6 +110,17 @@ def calc_solar_local_time(inst, lon_name=None, slt_name='slt'):
     if lon_name not in inst.data.keys():
         raise ValueError('unknown longitude variable name')
 
+    if ref_date is not None and apply_modulus:
+        estr = 'Keyword `ref_date` only supported if `apply_modulus`=False.'
+        raise ValueError(estr)
```
Comment (aburrell) [inline pysat/utils/coords.py:115]: I think this should at worst be a warning, though I would rank it at logging.info.  Supplying unused info isn't going to hurt anyone.

File: pysat/_files.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (aburrell) [reply pysat/_files.py]: Well, you have other comments that say the same thing down below.  If you found this helpful, then other people might too.  But in that case I think you should reword it.  Probably specify `ValueError: Can only compare identically-labeled Series objects` exactly and say that this is raised instead of False.  But, since you also say that down below where you do the ValueError catch, I think it could be removed here.

File: pysat/tests/test_orbits.py (hunk 9 above)
Comment (rstoneback) [reply pysat/tests/test_orbits.py:7]: It comes from the fun of  ```__repr__``` Everything needs to be just so. All the way at the end of the `__repr__`, Timedelta.
```
In [12]: test = pysat.Instrument('pysat', 'testing', 
    ...:                         clean_level='clean', 
    ...:                         orbit_info={'index': 'mlt', 'period':np.timedelta64(97, 'm')}, 
    ...:                         update_files=True)                                                                                                                                                                         

In [13]: test.orbits.__repr__()                                                                                                                                                                                             
Out[13]: "pysat.Orbits(inst=pysat.Instrument(platform='pysat', name='testing', inst_id='', clean_level='clean', pad=None, orbit_info={'index': 'mlt', 'period': numpy.timedelta64(97,'m')}, inst_module=None, custom=[], **{}), index='mlt', kind='local time', period=Timedelta('0 days 01:37:00'))"

In [14]: test.orbits.orbit_period                                                                                                                                                                                           
Out[14]: Timedelta('0 days 01:37:00')

In [15]: type(test.orbits.orbit_period)                                                                                                                                                                                     
Out[15]: pandas._libs.tslibs.timedeltas.Timedelta

```

```
In [20]: eval(test.orbits.__repr__())                                                                                                                                                                                       
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-20-b5dc860974c5> in <module>
----> 1 eval(test.orbits.__repr__())

<string> in <module>

NameError: name 'Timedelta' is not defined

```

```
In [21]: from pandas import Timedelta                                                                                                                                                                                       

In [22]: eval(test.orbits.__repr__())                                                                                                                                                                                       
Out[22]: pysat.Orbits(inst=pysat.Instrument(platform='pysat', name='testing', inst_id='', clean_level='clean', pad=None, orbit_info={'index': 'mlt', 'period': numpy.timedelta64(97,'m')}, inst_module=None, custom=[], **{}), index='mlt', kind='local time', period=Timedelta('0 days 01:37:00'))
```

File: pysat/_files.py [hunk 20]
```diff
@@ -266,6 +268,88 @@ def __str__(self):
...
+        -------
+        bool
+            True if objects are identical, False if they are not
+
+        """
+        # Check if the other object has the same type
+        if not isinstance(other, self.__class__):
+            return False
+
+        # If the type is the same CHECK FOR ALL THESE THINGS
```
[the remark this answers is not in the document]
Comment (rstoneback) [reply pysat/_files.py]: Sadly, I updated the text.

File: pysat/tests/test_orbits.py [hunk 21]
```diff
@@ -3,6 +3,8 @@
 import numpy as np
 
 import pandas as pds
+# Import below is needed for tests on __repr__
+from pandas import Timedelta # noqa
```
Comment (jklenzing) [inline pysat/tests/test_orbits.py:7]:
```suggestion
from pandas import Timedelta # noqa: F401
```
We should also specify which flake8 rule we're bypassing here.  It should be "module is imported but unused"

**Review by aburrell — COMMENTED** [review]
Getting there!  Nothing dire 🐺

File: pysat/tests/test_instrument.py [hunk 22]
```diff
@@ -843,6 +966,164 @@ def test_custom_instrument_load_3(self):
...
+                                                   'exit_night')
+                                                  ])
+    def test_instrument_partial_function_keywords(self, func, kwarg, val,
+                                                  caplog):
+        """Test if init keywords (partial funcs) are registered by pysat"""
+        # Test for file_date_range keyword
+        saved_level = pysat.logger.level
+        pysat.logger.setLevel(1)
+        caplog.set_level(logging.INFO)
+
```
Comment (jklenzing) [inline pysat/tests/test_instrument.py:1022]: See other caplog comment below

**Review by jklenzing — COMMENTED** [review]
There's a cleaner way of implementing the caplog usage.  However, we're not consistent elsewhere and this could be a separate issue for 3.0.1.

File: pysat/_instrument.py [hunk 23]
```diff
@@ -392,12 +401,15 @@ def __init__(self, platform=None, name=None, tag=None, inst_id=None,
                 saved_keys.extend(fkwargs)
 
                 # If the function can't access this dict, use partial
+                # The functions that truly can't access the dict are load,
```
[the remark this answers is not in the document]
Comment (rstoneback) [reply pysat/_instrument.py]: Thanks for being specific :)

I was looking at the code to come up with the proper NOUN but given my changes a couple comments had drifted out of date. I put in a minimum set of comment changes to best reflect the intent of your comment, that was the plan at least.

Given how this pull has gone this shouldn't be surprising, but I think I few lines of code in that method should actually come out now that things are more settled. I'm going to poke at that (very limited) and report back. I'll leave the option of what to do with it to y'all once I have the solution.

File: pysat/instruments/methods/testing.py [hunk 24]
```diff
@@ -1,50 +1,130 @@
...
                      "https://www.github.com/pysat/pysat"))
 
 # Load up citation information
-with NetworkLock(os.path.join(here, 'citation.txt'), 'r') as locked_file:
+with pysat.utils.NetworkLock(os.path.join(pysat.here, 'citation.txt'), 'r') as \
+        locked_file:
     refs = locked_file.read()
 
 
+def init(self, test_init_kwrd=None):
```
[the remark this answers is not in the document]
Comment (rstoneback) [reply pysat/instruments/methods/testing.py]: I found more instances of `kwrd`. Cleaned all that I found.

**Comment by aburrell** [comment]
> this could be a separate issue for 3.0.1.

YES!

**Comment by rstoneback** [comment]
> > this could be a separate issue for 3.0.1.
> 
> YES!

I missed a couple `kwrd` instances that were breaking tests so this small change made it through. Hands off other than the note above about potentially removing lines no longer really used.

**Comment by jklenzing** [comment]
Added #751 to address caplog usage

File: pysat/_instrument.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (rstoneback) [reply pysat/_instrument.py]: So the code in this area can be reduced to
```
        ...
        # Expected function keywords
        exp_keys = ['list_files', 'load', 'preprocess', 'download',
                    'list_remote_files', 'clean', 'init']
        for fkey in exp_keys:
            func_name = _kwargs_keys_to_func_name(fkey)
            func = getattr(self, func_name)

            # Get dict of supported keywords and values
            default_kwargs = _get_supported_keywords(func)

            # Check if kwargs are in list
            good_kwargs = [ckey for ckey in kwargs.keys()
                           if ckey in default_kwargs]

            # Store appropriate user supplied keywords for this function
            self.kwargs[fkey] = {gkey: kwargs[gkey] for gkey in good_kwargs}

            # Store all supported keywords for user edification
            self.kwargs_supported[fkey] = default_kwargs

            # Store keys to support check that all user supplied
            # keys are used.
            saved_keys.extend(default_kwargs.keys())

        # Test for user supplied keys that are not used
        ...
```

In the current pull attached functions (`_load_rtn`, `_list_files_rtn`, etc.) will always be partial functions. The changes above keep whatever function form is presented by the instrument support module. So for the testing instruments it can be a mixture of partial and regular functions. 

```
In [3]: test._load_rtn                                                                                                                                                                                                      
Out[3]: <function pysat.instruments.pysat_testing.load(fnames, tag=None, inst_id=None, sim_multi_file_right=False, sim_multi_file_left=False, root_date=None, malformed_index=False, num_samples=None, test_load_kwarg=None)>

In [4]: test._list_files_rtn                                                                                                                                                                                                
Out[4]: functools.partial(<function list_files at 0x7fb0d244fdc0>, test_dates={'': {'': datetime.datetime(2009, 1, 1, 0, 0), 'no_download': datetime.datetime(2009, 1, 1, 0, 0), 'non_strict': datetime.datetime(2009, 1, 1, 0, 0), 'user_password': datetime.datetime(2009, 1, 1, 0, 0), 'default_meta': datetime.datetime(2009, 1, 1, 0, 0)}})

```
 Per unit tests the end-level functionality is the same for both kinds of functions.

Technically there are options:
- [ ] Condense code, functions stay as they are given. This option is in code above as it is the minimum required. (unit tests pass)
- [ ] Keep the code as is in the pull now, everything is a partial function. 'Extra' lines do no harm. (unit tests pass)
- [ ] Condense code, make all functions partial functions (unit tests will pass)

How quickly things can spiral :) I can go for any of them really. Options 2 and 3 are more consistent than 1. That being said, the partialness of these functions isn't being directly used by pysat anymore with regards to keywords so no direct reason to make them all partial functions other than potential consistency.

[the remark this answers is not in the document]
Comment (rstoneback) [reply pysat/_instrument.py]: I should note that option 1 above is closest (?) to behavior that develop and the current actual RC branch have, mostly. Some instrument support functions are currently functions, others are partial. While some partial functions will always be partial, other functions remain how the instrument support module presents them.

So right now (develop-3) the `_*_rtn`s can be partial, or regular. The pull as is makes them all partial. Option 1 goes back to some are regular, some are partial.

[the remark this answers is not in the document]
Comment (rstoneback) [reply pysat/_instrument.py]: https://github.com/pysat/pysat/blob/4a2ed3a1f97ddbff5ae8c520ffe6d88b8ecf99a8/pysat/_instrument.py#L365

[the remark this answers is not in the document]
Comment (rstoneback) [reply pysat/_instrument.py]: I made #753 in an attempt to make this easier.

[the remark this answers is not in the document]
Comment (aburrell) [reply pysat/_instrument.py]: Since this has been captured in a separate PR, I believe this can be resolved.

File: pysat/_instrument.py [hunk 25]
```diff
@@ -3705,8 +3862,8 @@ def _get_supported_keywords(local_func):
     while len(func_args) > len(func_defaults):
         func_args.pop(0)
 
-    # remove pre-existing keywords from output
-    # first identify locations
+    # Remove pre-existing keywords from output
+    # First, identify locations
```
Comment (aburrell) [inline pysat/_instrument.py:3866]:
```suggestion
    # Remove pre-existing keywords from output. Start by identifying locations
```

**Review by aburrell — APPROVED** [review]
Small grammar correction, otherwise it looks good to me.

**Comment by rstoneback** [comment]
Thanks @aburrell and @jklenzing! I'm off to class prep, then class lecture, then post lecture recovery. Planning on getting the other pulls sorted out for another review cycle on Tues.

**Review by jklenzing — APPROVED** [review]
